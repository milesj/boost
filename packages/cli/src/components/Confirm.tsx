/* eslint-disable no-console */

import React, { useCallback, useEffect, useState } from 'react';
import { useFocus } from 'ink';
import { msg } from '../translate';
import { Prompt, PromptProps } from './internal/Prompt';
import { Selected } from './internal/Selected';
import { Style } from './Style';

export interface ConfirmProps extends Omit<PromptProps<boolean>, 'validate'> {
	/** Error message to display when an invalid character is pressed. Defaults to a custom message. */
	invalidError?: string;
	/** Character that triggers a falsy state when pressed. Defaults to "N". */
	no?: string;
	/** Character that triggers a truthy state when pressed. Defaults to "y". */
	yes?: string;
}

/**
 * A React component that renders a yes/no confirmation prompt.
 */
export function Confirm(props: ConfirmProps) {
	const { invalidError, onSubmit, no = 'N', yes = 'y', ...restProps } = props;
	const [value, setValue] = useState<boolean | null>(null);
	const { isFocused } = useFocus({ autoFocus: true });

	useEffect(() => {
		if (no.length !== 1) {
			console.error(`Prop "no" must be a single character.`);
		} else if (yes.length !== 1) {
			console.error(`Prop "yes" must be a single character.`);
		}
	}, [no, yes]);

	const handleInput = useCallback(
		(input: string) => {
			if (input === no) {
				setValue(false);
				onSubmit(false);
			} else if (input === yes) {
				setValue(true);
				onSubmit(true);
			} else {
				setValue(null);

				throw new Error(invalidError ?? msg('prompt:confirmInvalidValue', { no, yes }));
			}

			// Trigger submit
			return true;
		},
		[invalidError, no, onSubmit, yes],
	);

	return (
		<Prompt<boolean>
			{...restProps}
			afterLabel={
				value === null ? <Style type="muted">{`(${yes}/${no})`}</Style> : <Selected value={value} />
			}
			focused={isFocused}
			value={value}
			onInput={handleInput}
		/>
	);
}
