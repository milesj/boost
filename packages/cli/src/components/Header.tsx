import React from 'react';
import { Box } from 'ink';
import { SPACING_ROW } from '../constants';
import { StyleType } from '../types';
import { Style } from './Style';

export interface HeaderProps {
	/** Text to display for the label. */
	label: string;
	/** The bottom margin. Defaults to `1`. */
	marginTop?: number;
	/** The top margin. Defaults to `1`. */
	marginBottom?: number;
	/** Customize the background color using Style. Defaults to normal text. */
	type?: StyleType | 'none';
}

/**
 * A React component that renders a header/title within a terminal menu.
 */
export function Header(props: HeaderProps) {
	const { label, marginTop = SPACING_ROW, marginBottom = SPACING_ROW, type = 'default' } = props;

	return (
		<Box marginBottom={marginBottom} marginTop={marginTop}>
			<Style bold inverted type={type}>
				{` ${label.toLocaleUpperCase()} `}
			</Style>
		</Box>
	);
}
