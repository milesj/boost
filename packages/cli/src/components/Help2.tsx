import React from 'react';
import { Text } from 'ink';

export function Help() {
  return <Box flexDirection="column"></Box>;
}
