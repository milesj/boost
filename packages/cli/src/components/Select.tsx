import React, { useCallback, useMemo, useState } from 'react';
import { useFocus } from 'ink';
import { isObject } from '@boost/common';
import { figures } from '@boost/terminal';
import { useListNavigation } from '../hooks';
import { DividerRow } from './internal/DividerRow';
import { OptionRow } from './internal/OptionRow';
import { Prompt, PromptProps } from './internal/Prompt';
import { ScrollableList, ScrollableListProps } from './internal/ScrollableList';
import { Selected } from './internal/Selected';

export type SelectOptionLike<T> =
	| { divider: boolean; label?: NonNullable<React.ReactNode> }
	| { label: NonNullable<React.ReactNode>; value: T };

export interface SelectOption<T> {
	divider: boolean;
	index: number;
	label: NonNullable<React.ReactNode>;
	value: T | null;
}

export interface SelectProps<T, V = T> extends PromptProps<T>, ScrollableListProps {
	defaultSelected?: T;
	options: (SelectOptionLike<V> | V)[];
}

export function normalizeOptions<T>(options: SelectProps<unknown>['options']): SelectOption<T>[] {
	return options.map((option, index) => {
		if (isObject(option)) {
			return {
				divider: false,
				index,
				label: '',
				value: null,
				...(option as any),
			};
		}

		return {
			divider: false,
			index,
			label: String(option),
			value: option,
		};
	});
}

export function Select<T = string>({
	defaultSelected,
	limit,
	onSubmit,
	options: baseOptions,
	overflowAfterLabel,
	overflowBeforeLabel,
	scrollType,
	...props
}: SelectProps<T>) {
	const options = useMemo(() => normalizeOptions<T>(baseOptions), [baseOptions]);
	const [selectedValue, setSelectedValue] = useState<T | null>(defaultSelected ?? null);
	const { highlightedIndex, ...arrowKeyProps } = useListNavigation(options);
	const { isFocused } = useFocus({ autoFocus: true });

	const handleSpace = useCallback(() => {
		const { value } = options[highlightedIndex];

		if (value !== null) {
			if (value === selectedValue) {
				setSelectedValue(null);
			} else {
				setSelectedValue(value);
			}
		}
	}, [highlightedIndex, options, selectedValue]);

	const handleReturn = useCallback(() => {
		let value = selectedValue;

		if (value === null) {
			({ value } = options[highlightedIndex]);
			setSelectedValue(value);
		}

		// eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
		onSubmit?.(value!);

		// Trigger submit
		return true;
	}, [selectedValue, onSubmit, options, highlightedIndex]);

	return (
		<Prompt<T>
			{...props}
			{...arrowKeyProps}
			afterLabel={selectedValue !== null && <Selected value={selectedValue} />}
			focused={isFocused}
			value={selectedValue}
			onReturn={handleReturn}
			onSpace={handleSpace}
		>
			<ScrollableList
				currentIndex={highlightedIndex}
				items={options}
				limit={limit}
				overflowAfterLabel={overflowAfterLabel}
				overflowBeforeLabel={overflowBeforeLabel}
				renderItem={(option) => {
					if (option.divider) {
						return <DividerRow key={option.index} label={option.label} />;
					}

					return (
						<OptionRow
							key={option.index}
							highlighted={highlightedIndex === option.index}
							icon={figures.pointerSmall}
							iconActive={figures.pointer}
							label={option.label}
							selected={selectedValue === option.value}
						/>
					);
				}}
				scrollType={scrollType}
			/>
		</Prompt>
	);
}
