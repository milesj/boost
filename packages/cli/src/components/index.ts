export * from './Confirm';
export * from './Failure';
export * from './Header';
export * from './Help';
export * from './HiddenInput';
export * from './IndexHelp';
export * from './Input';
export * from './MultiSelect';
export * from './PasswordInput';
export * from './Select';
export * from './Style';
