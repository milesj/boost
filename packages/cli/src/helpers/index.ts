export * from './applyMarkdown';
export * from './applyStyle';
export * from './formatCommandCall';
export * from './formatDescription';
export * from './formatType';
export * from './formatValue';
export * from './getLongestWidth';
export * from './groupByCategory';
export * from './isArgvSize';
export * from './loadTheme';
export * from './mapCommandMetadata';
