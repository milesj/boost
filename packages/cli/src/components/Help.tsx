/* eslint-disable react/destructuring-assignment */

import React from 'react';
import { Box, Text } from 'ink';
import { OptionConfig, OptionConfigMap, ParamConfig, ParamConfigList } from '@boost/args';
import { toArray } from '@boost/common';
import { stripAnsi } from '@boost/terminal';
import { DELIMITER, SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../constants';
import {
	formatCommandCall,
	formatDescription,
	formatType,
	getLongestWidth,
	groupByCategory,
} from '../helpers';
import { msg } from '../translate';
import { Categories, CommandConfig, CommandConfigMap } from '../types';
import { Header } from './Header';
import { Style } from './Style';

export interface HelpProps {
	categories?: Categories;
	config?: CommandConfig;
	commands?: CommandConfigMap;
	delimiter?: string;
	header?: string;
	options?: OptionConfigMap;
	params?: ParamConfigList;
}

export class Help extends React.Component<HelpProps> {
	gatherOptionTags(config: OptionConfig): string[] {
		const tags: string[] = [];

		if (config.count) {
			tags.push(msg('cli:tagCount'));
		}

		if (config.multiple) {
			tags.push(msg('cli:tagMultiple') + (config.arity ? `(${config.arity})` : ''));
		}

		return tags;
	}

	gatherParamTags(config: ParamConfig): string[] {
		const tags: string[] = [];

		if (config.required) {
			tags.push(msg('cli:tagRequired'));
		}

		return tags;
	}

	renderCommands(commands: CommandConfigMap) {
		let pathWidth = 0;

		const items = Object.entries(commands).map(([path, command]) => {
			const pathCall = formatCommandCall(path, command);
			const pathCallStripped = stripAnsi(pathCall);

			if (pathCallStripped.length > pathWidth) {
				pathWidth = pathCallStripped.length;
			}

			return {
				...command,
				name: path,
				path,
				pathCall,
			};
		});

		const categorizedCommands = groupByCategory(items, this.props.categories ?? {});
		const categoryCount = Object.keys(categorizedCommands).length;

		return (
			<>
				<Header label={msg('cli:labelCommands')} />

				{Object.entries(categorizedCommands).map(([key, category], index) => (
					<Box key={key} flexDirection="column">
						{!!category.name && categoryCount > 1 && (
							<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
								<Style bold type="none">
									{category.name}
								</Style>
							</Box>
						)}

						{category.items.map((config) => {
							const desc = formatDescription(config);

							return (
								<Box key={key + config.path} flexDirection="row" paddingLeft={SPACING_COL}>
									<Box flexGrow={0} width={pathWidth + SPACING_COL_WIDE}>
										<Text>{config.pathCall}</Text>
									</Box>

									<Box flexGrow={1}>
										<Text wrap="wrap">{desc}</Text>
									</Box>
								</Box>
							);
						})}
					</Box>
				))}
			</>
		);
	}

	renderOptions(options: OptionConfigMap) {
		let longWidth = 0;
		let shortWidth = 0;

		const items = Object.entries(options).map(([name, option]) => {
			const shortLabel = option.short ? `-${option.short},` : '';
			let longLabel = `--${option.default === true ? 'no-' : ''}${name}`;

			if (option.type !== 'boolean') {
				longLabel += ' ';
				longLabel += formatType(option);
			}

			const longLabelStripped = stripAnsi(longLabel);
			const shortLabelStripped = stripAnsi(shortLabel);

			if (longLabelStripped.length > longWidth) {
				longWidth = longLabelStripped.length;
			}

			if (shortLabelStripped.length > shortWidth) {
				shortWidth = shortLabelStripped.length;
			}

			return {
				...option,
				longLabel,
				name,
				shortLabel,
			};
		});

		const categorizedOptions = groupByCategory(items, this.props.categories ?? {});
		const categoryCount = Object.keys(categorizedOptions).length;
		const showShortColumn = shortWidth > 0;

		return (
			<>
				<Header label={msg('cli:labelOptions')} />

				{Object.entries(categorizedOptions).map(([key, category], index) => (
					<Box key={key} flexDirection="column">
						{!!category.name && categoryCount > 1 && (
							<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
								<Style bold type="none">
									{category.name}
								</Style>
							</Box>
						)}

						{category.items.map((config) => (
							<Box key={key + config.name} flexDirection="row" paddingLeft={SPACING_COL}>
								{showShortColumn && (
									<Box flexGrow={0} width={shortWidth + SPACING_COL}>
										<Text>{config.shortLabel}</Text>
									</Box>
								)}

								<Box flexGrow={0} width={longWidth + SPACING_COL_WIDE}>
									<Text>{config.longLabel}</Text>
								</Box>

								<Box flexGrow={1}>
									<Text wrap="wrap">
										{formatDescription(config, this.gatherOptionTags(config))}
									</Text>
								</Box>
							</Box>
						))}
					</Box>
				))}
			</>
		);
	}

	renderParams(params: ParamConfigList) {
		const labels = params.map((config, index) => `${config.label ?? index} ${formatType(config)}`);
		const labelWidth = getLongestWidth(labels);
		const allowVariadic = this.props.config?.allowVariadicParams;

		return (
			<>
				<Header label={msg('cli:labelParams')} />

				{params.map((config, index) => {
					if (config.hidden) {
						return null;
					}

					const desc = config ? formatDescription(config, this.gatherParamTags(config)) : '';

					return (
						<Box key={`${labels[index]}-${index}`} flexDirection="row" paddingLeft={SPACING_COL}>
							<Box flexGrow={0} width={labelWidth + SPACING_COL_WIDE}>
								<Text>{labels[index]}</Text>
							</Box>

							<Box flexGrow={1}>
								<Text wrap="wrap">{desc}</Text>
							</Box>
						</Box>
					);
				})}

				{allowVariadic && (
					<Box paddingLeft={SPACING_COL}>
						<Text>
							{`â€¦${typeof allowVariadic === 'string' ? allowVariadic : ''}`}{' '}
							{formatType({
								multiple: true,
								type: 'string',
							})}
						</Text>
					</Box>
				)}
			</>
		);
	}

	renderUsage(usage: string[] | string = '') {
		const { delimiter = DELIMITER } = this.props;

		return (
			<>
				<Header label={msg('cli:labelUsage')} />

				{toArray(usage).map((example) => (
					<Box key={example} paddingLeft={SPACING_COL}>
						<Text>{example.startsWith(delimiter) ? example : delimiter + example}</Text>
					</Box>
				))}
			</>
		);
	}

	// eslint-disable-next-line complexity
	render() {
		const { commands, options, header, params, config } = this.props;
		const hasDesc = Boolean(config?.description);
		const hasUsage = Boolean(config?.usage && config.usage.length > 0);
		const hasParams = Boolean(params && params.length > 0);
		const hasCommands = Boolean(commands && Object.keys(commands).length > 0);
		const hasOptions = Boolean(options && Object.keys(options).length > 0);

		return (
			<Box flexDirection="column">
				{(hasDesc || hasUsage) && header && <Header label={header} />}

				{hasDesc && (
					<Box paddingLeft={SPACING_COL}>
						<Text>{formatDescription(config!)}</Text>
					</Box>
				)}

				{hasUsage && this.renderUsage(config?.usage)}

				{hasParams && this.renderParams(params!)}

				{hasCommands && this.renderCommands(commands!)}

				{hasOptions && this.renderOptions(options!)}
			</Box>
		);
	}
}
