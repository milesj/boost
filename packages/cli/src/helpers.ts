import { stripAnsi, style } from '@boost/terminal';
import { PrimitiveType } from '@boost/args/src';

export function applyMarkdown(text: string): string {
  return text.replace(/(?:\*|~|_){1,2}([^*~_]+)(?:\*|~|_){1,2}/giu, (match, body) => {
    if (match.startsWith('~')) {
      return style.strikethrough(body);
    } else if (match.startsWith('**') || match.startsWith('__')) {
      return style.bold(body);
    } else if (match.startsWith('*') || match.startsWith('_')) {
      return style.italic(body);
    }

    return match;
  });
}

export function formatValue(value: PrimitiveType): string {
  switch (typeof value) {
    case 'string':
      return `"${value}"`;
    default:
      return String(value);
  }
}

export function formatDescription(
  config: {
    choices?: PrimitiveType[];
    default?: PrimitiveType;
    description: string;
    deprecated?: boolean;
  },
  tags: string[] = [],
): string {
  let output = applyMarkdown(config.description);

  // Append choices after the description
  if (config.choices) {
    const choices = config.choices.map(
      choice => formatValue(choice) + (config.default === choice ? ' (default)' : ''),
    );

    output += ` (choices: ${choices.join(', ')})`;

    // Append default after description if no choices
  } else if (config.default) {
    output += ` (default: ${formatValue(config.default)})`;
  }

  // Tags go at the end of the description
  if (config.deprecated) {
    tags.unshift('deprecated');
  }

  if (tags.length > 0) {
    output += ' ';
    output += style.gray(`[${tags.join(', ')}]`);
  }

  return output;
}

export function formatType(
  config: {
    label?: string;
    multiple?: boolean;
    required?: boolean;
    type: string;
  },
  inline: boolean = false,
): string {
  let type = config.type + (config.multiple ? '[]' : '');

  if (inline) {
    if (config.label) {
      type = config.label;
    }

    type = config.required ? `<${type}>` : `[${type}]`;
  }

  return style.gray(type);
}

export function getLongestWidth(values: string[]): number {
  return values.reduce((sum, value) => {
    const text = stripAnsi(value);

    return text.length > sum ? text.length : sum;
  }, 0);
}
