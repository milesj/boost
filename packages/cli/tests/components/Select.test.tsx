import React from 'react';
import { render } from 'ink-testing-library';
import { Select, SelectProps } from '../../src/components/Select';
import { KEYS, options, optionsWithoutDivider } from '../helpers';

describe('Select', () => {
  const props: SelectProps<string> = {
    label: 'Favorite color?',
    options: [],
    onSubmit() {},
  };

  it('renders label by default', () => {
    const { lastFrame } = render(<Select {...props} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders options using strings', () => {
    const { lastFrame } = render(
      <Select {...props} options={optionsWithoutDivider.map((o) => o.value)} />,
    );

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders options using options', () => {
    const { lastFrame } = render(<Select {...props} options={optionsWithoutDivider} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders options and dividers using strings', () => {
    const { lastFrame } = render(
      <Select {...props} options={options.map((o) => ('divider' in o ? o : o.value))} />,
    );

    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders options and dividers using options', () => {
    const { lastFrame } = render(<Select {...props} options={options} />);

    expect(lastFrame()).toMatchSnapshot();
  });

  describe('limit', () => {
    describe('cycle scroll', () => {
      it('truncates list (odd)', () => {
        const { lastFrame } = render(
          <Select {...props} options={options} limit={5} scrollType="cycle" />,
        );

        expect(lastFrame()).toMatchSnapshot();
      });

      it('loops around when hitting an edge (odd)', async () => {
        const { lastFrame, stdin } = render(
          <Select {...props} options={options} limit={5} scrollType="cycle" />,
        );

        await delay();
        stdin.write(KEYS.up);
        stdin.write(KEYS.up);
        stdin.write(KEYS.up);
        await delay();

        expect(lastFrame()).toMatchSnapshot();
      });

      it('truncates list (even)', () => {
        const { lastFrame } = render(
          <Select {...props} options={options} limit={6} scrollType="cycle" />,
        );

        expect(lastFrame()).toMatchSnapshot();
      });

      it('loops around when hitting an edge (even)', async () => {
        const { lastFrame, stdin } = render(
          <Select {...props} options={options} limit={6} scrollType="cycle" />,
        );

        await delay();
        stdin.write(KEYS.up);
        stdin.write(KEYS.up);
        stdin.write(KEYS.up);
        stdin.write(KEYS.up);
        stdin.write(KEYS.up);
        await delay();

        expect(lastFrame()).toMatchSnapshot();
      });
    });

    describe('overflow scroll', () => {
      it('truncates list (odd)', () => {
        const { lastFrame } = render(
          <Select {...props} options={options} limit={5} scrollType="overflow" />,
        );

        expect(lastFrame()).toMatchSnapshot();
      });

      it('displays correct before/after more numbers (odd)', async () => {
        const { lastFrame, stdin } = render(
          <Select {...props} options={options} limit={5} scrollType="overflow" />,
        );

        await delay();
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        await delay();

        expect(lastFrame()).toMatchSnapshot();
      });

      it('truncates list (even)', () => {
        const { lastFrame } = render(
          <Select {...props} options={options} limit={6} scrollType="overflow" />,
        );

        expect(lastFrame()).toMatchSnapshot();
      });

      it('displays correct before/after more numbers (even)', async () => {
        const { lastFrame, stdin } = render(
          <Select {...props} options={options} limit={6} scrollType="overflow" />,
        );

        await delay();
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        stdin.write(KEYS.down);
        await delay();

        expect(lastFrame()).toMatchSnapshot();
      });
    });
  });

  describe('dividers', () => {
    describe('displays at the beginning', () => {
      const beginningDividerOptions = [{ divider: true }, 'one', 'two', 'three'];

      it('renders divider first and highlights first non-divider option', () => {
        const { lastFrame } = render(<Select {...props} options={beginningDividerOptions} />);

        expect(lastFrame()).toMatchSnapshot();
      });

      it('can cycle backwards past divider', async () => {
        const { lastFrame, stdin } = render(
          <Select {...props} options={beginningDividerOptions} />,
        );

        await delay();
        stdin.write(KEYS.up);
        await delay();

        expect(lastFrame()).toMatchSnapshot();
      });
    });

    describe('displays at the end', () => {
      const endDividerOptions = ['one', 'two', 'three', { divider: true }];

      it('renders divider last and highlights first non-divider option', () => {
        const { lastFrame } = render(<Select {...props} options={endDividerOptions} />);

        expect(lastFrame()).toMatchSnapshot();
      });

      it('can cycle forwards past divider', async () => {
        const { lastFrame, stdin } = render(<Select {...props} options={endDividerOptions} />);

        await delay();
        stdin.write(KEYS.down); // 2
        stdin.write(KEYS.down); // 3
        stdin.write(KEYS.down); // 1
        stdin.write(KEYS.down); // 2
        await delay();

        expect(lastFrame()).toMatchSnapshot();
      });
    });

    it('skips dividers while cycling until a value is found', async () => {
      const { lastFrame, stdin } = render(
        <Select
          {...props}
          options={[
            'one',
            { divider: true },
            { divider: true },
            { divider: true },
            { divider: true },
            'two',
          ]}
        />,
      );

      await delay();
      stdin.write(KEYS.down);
      await delay();

      expect(lastFrame()).toMatchSnapshot();
    });
  });

  describe('navigation', () => {
    const navOptions = ['one', 'two', 'three'];

    it('displays info color when highlighted', () => {
      const { lastFrame } = render(<Select {...props} options={navOptions} />);

      expect(lastFrame()).toMatchSnapshot();
    });

    it('displays notice color when selected', async () => {
      const { lastFrame, stdin } = render(<Select {...props} options={navOptions} />);

      await delay();
      stdin.write(KEYS.return);
      await delay();

      expect(lastFrame()).toMatchSnapshot();
    });

    it('can cycle forwards', async () => {
      const { lastFrame, stdin } = render(<Select {...props} options={navOptions} />);

      await delay();
      stdin.write(KEYS.down);
      await delay();

      expect(lastFrame()).toMatchSnapshot();
    });

    it('can cycle backwards', async () => {
      const { lastFrame, stdin } = render(<Select {...props} options={navOptions} />);

      await delay();
      stdin.write(KEYS.up);
      await delay();

      expect(lastFrame()).toMatchSnapshot();
    });

    it('can jump between last and first', async () => {
      const { lastFrame, stdin } = render(<Select {...props} options={navOptions} />);

      await delay();
      stdin.write(KEYS.right);
      await delay();

      expect(lastFrame()).toMatchSnapshot();

      stdin.write(KEYS.left);
      await delay();

      expect(lastFrame()).toMatchSnapshot();
    });
  });

  describe('selection', () => {
    it('does nothing when hitting space bar', async () => {
      const spy = jest.fn();
      const { lastFrame, stdin } = render(
        <Select {...props} limit={5} options={options} onSubmit={spy} />,
      );

      await delay();
      stdin.write(' ');
      await delay();

      expect(lastFrame()).toMatchSnapshot();
      expect(spy).not.toHaveBeenCalled();
    });

    it('calls `onSubmit` with selected value when hitting enter', async () => {
      const spy = jest.fn();
      const { lastFrame, stdin } = render(
        <Select {...props} limit={5} options={options} onSubmit={spy} />,
      );

      await delay();
      stdin.write(KEYS.return);
      await delay();

      expect(lastFrame()).toMatchSnapshot();
      expect(spy).toHaveBeenCalledWith('black');
    });
  });
});
