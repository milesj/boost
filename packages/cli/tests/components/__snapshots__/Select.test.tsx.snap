// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select dividers displays at the beginning can cycle backwards past divider 1`] = `
"[36m?[39m [1mFavorite color?[22m
    [90m‚îÄ‚îÄ‚îÄ‚îÄ[39m
  [90m‚Ä∫[39m one
  [90m‚Ä∫[39m two
  [36m‚ùØ[39m [36mthree[39m"
`;

exports[`Select dividers displays at the beginning renders divider first and highlights first non-divider option 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select dividers displays at the end can cycle forwards past divider 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [90m‚Ä∫[39m one
  [36m‚ùØ[39m [36mtwo[39m
  [90m‚Ä∫[39m three
    [90m‚îÄ‚îÄ‚îÄ‚îÄ[39m"
`;

exports[`Select dividers displays at the end renders divider last and highlights first non-divider option 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select dividers skips dividers while cycling until a value is found 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [90m‚Ä∫[39m one
    [90m‚îÄ‚îÄ‚îÄ‚îÄ[39m
    [90m‚îÄ‚îÄ‚îÄ‚îÄ[39m
    [90m‚îÄ‚îÄ‚îÄ‚îÄ[39m
    [90m‚îÄ‚îÄ‚îÄ‚îÄ[39m
  [36m‚ùØ[39m [36mtwo[39m"
`;

exports[`Select limit cycle scroll loops around when hitting an edge (even) 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [36m‚ùØ[39m [36mOrange[39m
    [90mP[39m
  [90m‚Ä∫[39m Purple
    [90mR[39m
  [90m‚Ä∫[39m Red
    [90mW[39m"
`;

exports[`Select limit cycle scroll loops around when hitting an edge (odd) 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [36m‚ùØ[39m [36mRed[39m
    [90mW[39m
  [90m‚Ä∫[39m White
    [90mY[39m
  [90m‚Ä∫[39m Yellow"
`;

exports[`Select limit cycle scroll truncates list (even) 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select limit cycle scroll truncates list (odd) 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [36m‚ùØ[39m [36mBlack[39m
  [90m‚Ä∫[39m Blue
  [90m‚Ä∫[39m Brown
    [90mC[39m
  [90m‚Ä∫[39m Cyan"
`;

exports[`Select limit overflow scroll displays correct before/after more numbers (even) 1`] = `
"[36m?[39m [1mFavorite color?[22m
    [90m3 more[39m
  [90m‚Ä∫[39m Cyan
    [90mG[39m
  [90m‚Ä∫[39m Gray
  [36m‚ùØ[39m [36mGreen[39m
    [90mO[39m
  [90m‚Ä∫[39m Orange
    [90m4 more[39m"
`;

exports[`Select limit overflow scroll displays correct before/after more numbers (odd) 1`] = `
"[36m?[39m [1mFavorite color?[22m
    [90m4 more[39m
    [90mG[39m
  [90m‚Ä∫[39m Gray
  [36m‚ùØ[39m [36mGreen[39m
    [90mO[39m
  [90m‚Ä∫[39m Orange
    [90m4 more[39m"
`;

exports[`Select limit overflow scroll truncates list (even) 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select limit overflow scroll truncates list (odd) 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select navigation can cycle backwards 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [90m‚Ä∫[39m one
  [90m‚Ä∫[39m two
  [36m‚ùØ[39m [36mthree[39m"
`;

exports[`Select navigation can cycle forwards 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [90m‚Ä∫[39m one
  [36m‚ùØ[39m [36mtwo[39m
  [90m‚Ä∫[39m three"
`;

exports[`Select navigation can jump between last and first 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [90m‚Ä∫[39m one
  [90m‚Ä∫[39m two
  [36m‚ùØ[39m [36mthree[39m"
`;

exports[`Select navigation can jump between last and first 2`] = `
"[36m?[39m [1mFavorite color?[22m
  [36m‚ùØ[39m [36mone[39m
  [90m‚Ä∫[39m two
  [90m‚Ä∫[39m three"
`;

exports[`Select navigation displays info color when highlighted 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select navigation displays notice color when selected 1`] = `
"[32m‚úî[39m [1mFavorite color?[22m [35mone[39m
  [36m‚ùØ[39m [36mone[39m
  [90m‚Ä∫[39m two
  [90m‚Ä∫[39m three"
`;

exports[`Select renders label by default 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select renders options and dividers using options 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select renders options and dividers using strings 1`] = `
"[36m?[39m [1mFavorite color?[22m
    [90mB[39m
  [36m‚ùØ[39m [36mblack[39m
  [90m‚Ä∫[39m blue
  [90m‚Ä∫[39m brown
    [90mC[39m
  [90m‚Ä∫[39m cyan
    [90mG[39m
  [90m‚Ä∫[39m gray
  [90m‚Ä∫[39m green
    [90mO[39m
  [90m‚Ä∫[39m orange
    [90mP[39m
  [90m‚Ä∫[39m purple
    [90mR[39m
  [90m‚Ä∫[39m red
    [90mW[39m
  [90m‚Ä∫[39m white
    [90mY[39m
  [90m‚Ä∫[39m yellow"
`;

exports[`Select renders options using options 1`] = `"[36m?[39m [1mFavorite color?[22m"`;

exports[`Select renders options using strings 1`] = `
"[36m?[39m [1mFavorite color?[22m
  [36m‚ùØ[39m [36mblack[39m
  [90m‚Ä∫[39m blue
  [90m‚Ä∫[39m brown
  [90m‚Ä∫[39m cyan
  [90m‚Ä∫[39m gray
  [90m‚Ä∫[39m green
  [90m‚Ä∫[39m orange
  [90m‚Ä∫[39m purple
  [90m‚Ä∫[39m red
  [90m‚Ä∫[39m white
  [90m‚Ä∫[39m yellow"
`;

exports[`Select selection calls \`onSubmit\` with highlighted (but not selected) value when hitting enter 1`] = `
"[32m‚úî[39m [1mFavorite color?[22m [35mblack[39m
    [90mB[39m
  [36m‚ùØ[39m [36mBlack[39m
  [90m‚Ä∫[39m Blue
  [90m‚Ä∫[39m Brown
    [90mC[39m
    [90m8 more[39m"
`;

exports[`Select selection calls \`onSubmit\` with selected value when hitting enter 1`] = `
"[32m‚úî[39m [1mFavorite color?[22m [35mblack[39m
    [90m2 more[39m
  [90m‚Ä∫[39m Brown
    [90mC[39m
  [36m‚ùØ[39m [36mCyan[39m
    [90mG[39m
  [90m‚Ä∫[39m Gray
    [90m6 more[39m"
`;

exports[`Select selection can unselect a value when hitting space bar again 1`] = `
"[36m?[39m [1mFavorite color?[22m [35mblack[39m
    [90mB[39m
  [35m‚ùØ[39m [35mBlack[39m
  [36m‚ùØ[39m [36mBlue[39m
  [90m‚Ä∫[39m Brown
    [90mC[39m
    [90m8 more[39m"
`;

exports[`Select selection can unselect a value when hitting space bar again 2`] = `
"[36m?[39m [1mFavorite color?[22m [35mblue[39m
    [90mB[39m
  [90m‚Ä∫[39m Black
  [36m‚ùØ[39m [36mBlue[39m
  [90m‚Ä∫[39m Brown
    [90mC[39m
    [90m8 more[39m"
`;

exports[`Select selection selects and displays a value when hitting space bar 1`] = `
"[36m?[39m [1mFavorite color?[22m [35mblack[39m
    [90mB[39m
  [35m‚ùØ[39m [35mBlack[39m
  [36m‚ùØ[39m [36mBlue[39m
  [90m‚Ä∫[39m Brown
    [90mC[39m
    [90m8 more[39m"
`;
